specs/_features/whisk/beacon-chain.md

```diff
@@ -0,0 +1,490 @@
+    amount: uint64
+) -> Validator:
+    effective_balance = min(amount - amount % EFFECTIVE_BALANCE_INCREMENT, MAX_EFFECTIVE_BALANCE)
+    k = get_unique_whisk_k(state, ValidatorIndex(len(state.validators)))
+    whisk_k_commitment, whisk_tracker = get_initial_commitments(k)
+
+    validator = Validator(
+        pubkey=pubkey,
+        withdrawal_credentials=withdrawal_credentials,
+        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+        # Whisk fields
+        whisk_tracker=whisk_tracker,
+        whisk_k_commitment=whisk_k_commitment,
+    )
+    return validator
+```
+
+```python
+def process_deposit(state: BeaconState, deposit: Deposit) -> None:
+    # Verify the Merkle branch
+    assert is_valid_merkle_branch(
+        leaf=hash_tree_root(deposit.data),
+        branch=deposit.proof,
+        depth=DEPOSIT_CONTRACT_TREE_DEPTH + 1,  # Add 1 for the List length mix-in
+        index=state.eth1_deposit_index,
+        root=state.eth1_data.deposit_root,
+    )
+
+    # Deposits must be processed in order
+    state.eth1_deposit_index += 1
+
+    pubkey = deposit.data.pubkey
+    amount = deposit.data.amount
+    validator_pubkeys = [validator.pubkey for validator in state.validators]
+    if pubkey not in validator_pubkeys:
+        # Verify the deposit signature (proof of possession) which is not checked by the deposit contract
+        deposit_message = DepositMessage(
+            pubkey=deposit.data.pubkey,
+            withdrawal_credentials=deposit.data.withdrawal_credentials,
+            amount=deposit.data.amount,
+        )
+        domain = compute_domain(DOMAIN_DEPOSIT)  # Fork-agnostic domain since deposits are valid across forks
+        signing_root = compute_signing_root(deposit_message, domain)
+        # Initialize validator if the deposit signature is valid
+        if bls.Verify(pubkey, signing_root, deposit.data.signature):
+            state.validators.append(
+                get_validator_from_deposit_whisk(state, pubkey, deposit.data.withdrawal_credentials, amount)
+            )  # [Change in Whisk]
+            state.balances.append(amount)
+            state.previous_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.current_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.inactivity_scores.append(uint64(0))
+    else:
+        # Increase balance by deposit amount
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        increase_balance(state, index, amount)
+```
+
+### `get_beacon_proposer_index`
+
+```python
+def get_beacon_proposer_index(state: BeaconState) -> ValidatorIndex:
+    """
+    Return the beacon proposer index at the current slot.
+    """
+    print("get_beacon_proposer_index", state.latest_block_header.slot, state.slot)
```

Comment:
Hmm this function was slightly changed and a print was added. Is this on purpose?

Reply:
Will remove, comes from the original PR so unsure of the purpose. Maybe debugging artifacts from attempts to test https://github.com/ethereum/consensus-specs/pull/3205/files#diff-a8e88205c050488d796a3a2240bc71338a325bc396fd241b4f40b21b82a8181cR361-R371

Reply:
Oh I see, sorry for the confusion. Tho if feels a bit like a hack it's better than changing code everywhere. As `state.latest_block_header` is updated before any other action in `process_block`, it's looks good to me